src/network/networkd-dhcp4.c

```diff
@@ -462,12 +462,19 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
         }
 
         if (link->network->dhcp_use_hostname) {
-                const char *hostname = NULL;
+                const char *dhcpname = NULL;
+                _cleanup_free_ char *hostname = NULL;
 
                 if (link->network->dhcp_hostname)
-                        hostname = link->network->dhcp_hostname;
+                        dhcpname = link->network->dhcp_hostname;
                 else
-                        (void) sd_dhcp_lease_get_hostname(lease, &hostname);
+                        (void) sd_dhcp_lease_get_hostname(lease, &dhcpname);
+
+                if (dhcpname) {
+                        r = shorten_overlong(dhcpname, &hostname);
+                        if (r < 0)
+                                log_link_error_errno(link, r, "Failed to shorten DHCP hostname '%s': %m", dhcpname);
```

Comment:
i think we should log here about the shortening. if you make shorten_overlong() return == 0 and > 0 as suggested above we could add a notice message here saying "DHCP supplied hostname too long for Linux, shortened from %s to %s" or so...

Also, the error case should be downgraded to warning really, as we proceed otherwise, the issue isn't fatal. And we should indicate that, maybe by suffixing the error message with "Unable to shorten DHCP supplied hostname '%s', ignoring: %m"